HPCC-18406 Use global and local ids to track related transactions

afishbeck opened the pull request:
In order to create a giant graph of every system call related to a
global product order/request, an external system may pass a globalId
and a "callerId" to an HPCC component.

GlobalIds tie together the entire customer order/request across
all components.

CallerIds tie together all the calls made from a particular transaction
node.

When globalIds are recieved we should:

1. Log the global and caller ids that are received.
2. Create our own local id representing the specific request on
   the current node.
2. When making any SOAPCALL or HTTPCALL automatically pass along the
   globalId and pass our localId as the outgoing "callerId".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 3611ed1bea997b122ab8a541bf631c7a7b8c3289
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |727 |726 |1 |
|test (thor) |640 |639 |1 |
|test (roxie) |755 |754 |1 |


[Errors:](https://gist.github.com/3ded3eea41e3881ad10a7a27c548ad40)
- test
  - hthor
    - 274\. [httpcall_multiheader](https://gist.github.com/HPCCSmoketest/3ded3eea41e3881ad10a7a27c548ad40#file-httpcall_multiheader-httpcall_multiheader-ecl-log), 
  - thor
    - 255\. [httpcall_multiheader](https://gist.github.com/HPCCSmoketest/3ded3eea41e3881ad10a7a27c548ad40#file-httpcall_multiheader-httpcall_multiheader-ecl-log), 
  - roxie
    - 272\. [httpcall_multiheader](https://gist.github.com/HPCCSmoketest/3ded3eea41e3881ad10a7a27c548ad40#file-httpcall_multiheader-httpcall_multiheader-ecl-log), 

HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@afishbeck It seems your changes alter the test output, but the keyfile didn't changed and that's why httpcall_multiheader.ecl failed.

richardkchapman commented on common/thorhelper/thorsoapcall.cpp:
```diff
@@ -1524,6 +1524,12 @@ class CWSCAsyncFor : implements IWSCAsyncFor, public CInterface, public CAsyncFo
 #endif
+        if (!isEmptyString(master->logctx.queryGlobalId()))
+        {
+            request.append("RELX-Global-Id: ").append(master->logctx.queryGlobalId()).append("\r\n");
```
This feels rather RELX specific for OSS code. Is there a reason the Global ID would only be useful to RELX?

afishbeck replied:
It's more that all RISK components will be using the same headers.  I.e. web applications, scapps.  Now is the time to set the standard, we could do HPCC, but not sure what would fit for all of RISK (and perhaps eventually beyond).

richardkchapman replied:
What about users of HPCC outside of RELX ? Why would they want to use a global id called "RELX-Global-Id" ?

afishbeck replied:
Perhaps something even more generic.  I thought just Global-ID might not be unique enough.

afishbeck replied:
I'll bring up a more generic name with the Global ID team.

richardkchapman replied:
Alternatively, make the name configurable.

afishbeck replied:
Yes, I'll do that.

afishbeck commented:
@richardkchapman note the inclusion of libbase58 code which is MIT license.  I'm not sure what the current practice is.  For now simply adding it was simplest.  MIT licesnse is very flexible.  base58 was chosen for the IDs (and already in use in RISK india) because it's human readable.  I.e. a customer could read the number to a customer support person.

afishbeck commented:
@AttilaVamos right, thanks, I'll add the update for the httpcall_multiheader.ecl key.

AttilaVamos commented:
@afishbeck back to you.

richardkchapman commented:
Needs rebasing for a clean merge

afishbeck committed:
HPCC-18406 Use global and local ids to track related transactions

In order to create a giant graph of every system call related to a
global product order/request, an external system may pass a globalId
and a "callerId" to an HPCC component.

GlobalIds tie together the entire customer order/request across
all components.

CallerIds tie together all the calls made from a particular transaction
node.

When globalIds are recieved we should:

1. Log the global and caller ids that are received.
2. Create our own local id representing the specific request on
   the current node.
2. When making any SOAPCALL or HTTPCALL automatically pass along the
   globalId and pass our localId as the outgoing "callerId".

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@richardkchapman updated to make it more generic for external users, and made it configurable so we could match the internal HTTP header name.

Also rebased.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: c23e3796520ee8b9ccebec0ea60d204affd84198
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |0 |0 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |734 |734 |0 |
|test (thor) |647 |647 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
